File: app/src/main/java/com/pkmnapps/activitydo/adapters/ListAdapter.java
package com.pkmnapps.activitydo.adapters;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageButton;

import com.pkmnapps.activitydo.R;

import com.pkmnapps.activitydo.custominterfaces.ListActivityInterface;
import com.pkmnapps.activitydo.databasehelpers.DBHelperListItems;
import com.pkmnapps.activitydo.dataclasses.ListItem;

import java.util.List;

public class ListAdapter extends RecyclerView.Adapter<ListAdapter.MyViewHolder> {

    private final List<ListItem> listItems;
    final ListActivityInterface listActivityInterface;

    public static class MyViewHolder extends RecyclerView.ViewHolder {
        public final EditText content;
        public final CheckBox checked;
        public final ImageButton delete;
        public MyViewHolder(View view) {
            super(view);
            content = view.findViewById(R.id.contentEditText);
            checked = view.findViewById(R.id.checkbox);
            delete = view.findViewById(R.id.deleteButton);
        }
    }

    public ListAdapter(List<ListItem> listItems, ListActivityInterface listActivityInterface) {
        this.listItems = listItems;
        this.listActivityInterface = listActivityInterface;
    }

    @NonNull
    @Override
    public ListAdapter.MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.recycler_listitem_layout, parent, false);

        return new MyViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull final ListAdapter.MyViewHolder holder, final int position) {
        final ListItem listItem = listItems.get(position);

        holder.content.setText(listItem.getContent());
        holder.checked.setChecked(listItem.getChecked());
        holder.checked.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listItem.setChecked(holder.checked.isChecked());
            }
        });
        holder.delete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //remove from database if exists
                new DBHelperListItems(holder.itemView.getContext()).deleteListItem(listItem.getUid());
                listItems.remove(listItem);
                listActivityInterface.deleteListItem(listItem);
            }
        });
        holder.content.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if(!hasFocus) {
                    listItem.setContent(holder.content.getText().toString());
                    holder.delete.setVisibility(View.INVISIBLE);
                }
                else {//gained focus
                    holder.delete.setVisibility(View.VISIBLE);
                }
            }
        });
        holder.content.requestFocus();
    }

    @Override
    public int getItemCount() {
        return listItems.size();
    }

}

File: app/src/main/java/com/pkmnapps/activitydo/adapters/ListAdapter.kt
package com.pkmnapps.activitydo.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.EditText
import android.widget.ImageButton
import androidx.recyclerview.widget.RecyclerView
import com.pkmnapps.activitydo.R
import com.pkmnapps.activitydo.custominterfaces.ListActivityInterface
import com.pkmnapps.activitydo.databasehelpers.DBHelperListItems
import com.pkmnapps.activitydo.dataclasses.ListItem
import org.junit.runner.RunWith

class ListAdapter(private val listItems: MutableList<ListItem?>?, val listActivityInterface: ListActivityInterface?) : RecyclerView.Adapter<ListAdapter.MyViewHolder?>() {
    class MyViewHolder(view: View?) : RecyclerView.ViewHolder(view) {
        val content: EditText?
        val checked: CheckBox?
        val delete: ImageButton?

        init {
            content = view.findViewById(R.id.contentEditText)
            checked = view.findViewById(R.id.checkbox)
            delete = view.findViewById(R.id.deleteButton)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val itemView = LayoutInflater.from(parent.context)
                .inflate(R.layout.recycler_listitem_layout, parent, false)
        return MyViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val listItem = listItems.get(position)
        holder.content.setText(listItem.getContent())
        holder.checked.setChecked(listItem.getChecked())
        holder.checked.setOnClickListener(View.OnClickListener { listItem.setChecked(holder.checked.isChecked()) })
        holder.delete.setOnClickListener(View.OnClickListener { //remove from database if exists
            DBHelperListItems(holder.itemView.context).deleteListItem(listItem.getUid())
            listItems.remove(listItem)
            listActivityInterface.deleteListItem(listItem)
        })
        holder.content.setOnFocusChangeListener(OnFocusChangeListener { v, hasFocus ->
            if (!hasFocus) {
                listItem.setContent(holder.content.getText().toString())
                holder.delete.setVisibility(View.INVISIBLE)
            } else { //gained focus
                holder.delete.setVisibility(View.VISIBLE)
            }
        })
        holder.content.requestFocus()
    }

    override fun getItemCount(): Int {
        return listItems.size
    }
}